sql/binlog.cc

```diff
@@ -9146,12 +9156,127 @@ static int binlog_start_trans_and_stmt(THD *thd, Log_event *start_event) {
 
     Query_log_event qinfo(thd, query, qlen, is_transactional, false, true, 0,
                           true);
-    if (cache_data->write_event(&qinfo)) DBUG_RETURN(1);
+    if (cache_data->write_event(thd, &qinfo)) DBUG_RETURN(1);
   }
 
   DBUG_RETURN(0);
 }
 
+/**
+  This function generated meta data in JSON format as a comment in a rows query
+  event.
+  @see binlog_trx_meta_data
+  @return JSON if all good, null string otherwise
+*/
+std::string THD::gen_trx_metadata() {
+  DBUG_ENTER("THD::gen_trx_metadata");
+  DBUG_ASSERT(opt_binlog_trx_meta_data);
+
+  ptree pt;
+  std::ostringstream buf;
+
+  // case: read existing meta data received from the master
+  if (rli_slave && !rli_slave->trx_meta_data_json.empty()) {
+    std::istringstream is(rli_slave->trx_meta_data_json);
+    try {
+      read_json(is, pt);
+    } catch (const std::exception &e) {
+      // NO_LINT_DEBUG
+      sql_print_error("Exception while reading meta data: %s, JSON: %s",
+                      e.what(), rli_slave->trx_meta_data_json.c_str());
+      DBUG_RETURN("");
+    }
+    // clear existing data
+    rli_slave->trx_meta_data_json.clear();
+  }
+
+  // add things to the meta data
+  try {
+    add_time_metadata(pt);
+    // uncomment after db_metadata is implemented
+    // add_db_metadata(pt);
+  } catch (const std::exception &e) {
+    // NO_LINT_DEBUG
+    sql_print_error("Exception while adding meta data: %s", e.what());
+    DBUG_RETURN("");
+  }
+
+  // write meta data with new stuff in the binlog
+  try {
+    write_json(buf, pt, false);
+  } catch (const std::exception &e) {
+    // NO_LINT_DEBUG
+    sql_print_error("Exception while writing meta data: %s", e.what());
+    DBUG_RETURN("");
+  }
+  std::string json = buf.str();
+  boost::trim(json);
+
+  std::string comment_str =
+      std::string("/*").append(TRX_META_DATA_HEADER).append(json).append("*/");
+
+  DBUG_RETURN(comment_str);
+}
+
+/**
+  This function adds timing information in meta data JSON of rows query event.
+  @see THD::write_trx_metadata
+  @param meta_data_root Property tree object which represents the JSON
+*/
+void THD::add_time_metadata(ptree &meta_data_root) {
+  DBUG_ENTER("THD::add_time_metadata");
+  DBUG_ASSERT(opt_binlog_trx_meta_data);
+
+  // get existing timestamps
+  ptree timestamps = meta_data_root.get_child("ts", ptree());
```

Comment:
You can capture the result by reference here to avoid
```c++
  // update timestamps in root
  meta_data_root.erase("ts");
  meta_data_root.add_child("ts", timestamps);
```
```suggestion
  ptree& timestamps = meta_data_root.get_child("ts", ptree());
```

Reply:
Not possible: this `get_child` overload returns a const reference.